api/service/service.go

```diff
@@ -0,0 +1,128 @@
+package service
+
+import (
+	"context"
+	uuid "github.com/satori/go.uuid"
+	"container/list"
+	"fmt"
+	"io"
+	"encoding/json"
+
+	"github.com/spaceuptech/space-api-go/api/proto"
+	"github.com/spaceuptech/space-api-go/api/config"
+	"github.com/spaceuptech/space-api-go/api/model"
+	"github.com/spaceuptech/space-api-go/api/utils"
+)
+
+type CallBackFunction func(string, interface{})()
+type ServiceFunction func(*model.Message, *model.Message, CallBackFunction)()
+
+// Service contains the methods for the service instance
+type Service struct {
+	config  *config.Config
+	service string
+	id      string
+	funcs   map[string]ServiceFunction
+}
+
+func Init(config *config.Config, service string) *Service {
+	id := uuid.NewV1().String()
+	return &Service{config, service, id, make(map[string]ServiceFunction)}
+}
+
+func (s *Service) RegisterFunc(funcName string, function ServiceFunction) {
+	s.funcs[funcName] = function
+}
+
+func (s *Service) Start() {
+	registerRequest := &proto.FunctionsPayload{Service: s.service, Type: utils.TypeServiceRegister, Id: s.id, Project: s.config.Project, Token: s.config.Token}
+	l := list.New()
+	conn := s.config.Transport.GetConn()
+	ctx, cancel := context.WithCancel(context.Background())
+	var toClose bool = false
```

Comment:
Remove the toClose mechanism. Just return from start.

Reply:
I think that is required. If we get an EOF, we exit from start. But if we get some other error, we need to exit from the inner function and try again

Reply:
No need to have the inner function